File: haskell/test/PointTests.hs
import Test.HUnit
import Test.Framework
import Test.Framework.Providers.HUnit
import Data.Monoid
import Control.Monad
import PointData
import Utility

-- Data
startPoint :: Point2D
startPoint = Point 0.0 0.0

-- dist Tests
distError :: String
distError = "Distance was not calculated correctly"

distTest1 :: Assertion
distTest2 :: Assertion

distTest1 = assertEqual distError 0.0 (dist (Point 1.0 1.0) (Point 1.0 1.0))
distTest2 = assertEqual distError 5.0 (dist (Point 3.0 0.0) (Point 0.0 4.0))

-- startingPoint Tests
startError :: String
startError = "Did not find the correct start point"

startTest1 :: Assertion
startTest2 :: Assertion

startTest1 = assertEqual startError startPoint (startingPoint [startPoint])
startTest2 = assertEqual startError (Point 1.0 1.0) (startingPoint [(Point 2.0 2.0), (Point 2.0 1.0), (Point 1.0 1.0), (Point 1.0 2.0)])

-- angleBetween Tests
polarError :: String
polarError = "Polar angle was not calculated correctly"

polarTest :: Point2D -> Double -> Assertion
polarTest p expected = assertBool polarError (approxEqual expected (angleBetween startPoint p))

polarNotEqual :: String
polarNotEqual = "Angles not equal"

haveSameAngle :: Point2D -> Point2D -> Assertion
haveSameAngle a b = assertBool polarNotEqual (approxEqual (angleBetween startPoint a) (angleBetween startPoint b))

polarTest0Deg   :: Assertion
polarTest30Deg  :: Assertion
polarTest60Deg  :: Assertion
polarTest90Deg  :: Assertion
polarTest120Deg :: Assertion
polarTest150Deg :: Assertion
polarTest180Deg :: Assertion

polarTest0Deg   = polarTest (Point 1.0 0.0) 0.0
polarTest30Deg  = polarTest (Point ((sqrt 3) / 2.0) 0.5) (pi / 6.0)
polarTest60Deg  = polarTest (Point 0.5 ((sqrt 3) / 2.0)) (pi / 3.0)
polarTest90Deg  = polarTest (Point 0.0 1.0) (pi / 2.0)
polarTest120Deg = polarTest (Point (negate 0.5) ((sqrt 3) / 2.0)) (4.0 * pi / 6.0)
polarTest150Deg = polarTest (Point (negate ((sqrt 3) / 2.0)) 0.5) (5.0 * pi / 6.0)
polarTest180Deg = polarTest (Point (negate 1.0) 0.0) pi

polarTestLine = haveSameAngle (Point 1.0 2.0) (Point 2.0 4.0)

-- dedupByEq Tests
dedupError :: String
dedupError = "Did not deduplicate list correctly"

dedupByEqTest :: (Eq a, Show a) => [a] -> [a] -> Assertion
dedupByEqTest ps expected = assertEqual dedupError expected (dedupByEq ps)

dedupByEqNone   :: Assertion
dedupByEqOne    :: Assertion
dedupByEqTriple :: Assertion
dedupByEqTwo    :: Assertion

dedupByEqNone   = dedupByEqTest [(Point 1.0 2.0), (Point 1.0 1.0)]
                                [(Point 1.0 2.0), (Point 1.0 1.0)]
dedupByEqOne    = dedupByEqTest [(Point 1.0 1.0), (Point 1.0 1.0)]
                                [(Point 1.0 1.0)]
dedupByEqTriple = dedupByEqTest [(Point 1.0 1.0), (Point 1.0 1.0), (Point 1.0 1.0)]
                                [(Point 1.0 1.0)]
dedupByEqTwo    = dedupByEqTest [(Point 1.0 1.0), (Point 2.0 3.0), (Point 2.0 3.0), (Point 1.0 1.0)]
                                [(Point 1.0 1.0), (Point 2.0 3.0)]

-- dedupByAngle Tests
dedupByAngleTest :: [Point2D] -> [Point2D] -> Assertion
dedupByAngleTest ps expected = assertEqual dedupError expected (dedupByAngle startPoint ps)

dedupByAngleNone    :: Assertion
dedupByAngleOne     :: Assertion
dedupByAngleTriple  :: Assertion
dedupByAngleTwoA    :: Assertion
dedupByAngleTwoB    :: Assertion

dedupByAngleNone    = dedupByAngleTest  [(Point 1.0 2.0), (Point 1.0 1.0)]
                                        [(Point 1.0 2.0), (Point 1.0 1.0)]
dedupByAngleOne     = dedupByAngleTest  [(Point 1.0 2.0), (Point 10.0 20.0)]
                                        [(Point 10.0 20.0)]
dedupByAngleTriple  = dedupByAngleTest  [(Point 1.0 2.0), (Point 10.0 20.0), (Point 100.0 200)]
                                        [(Point 100.0 200.0)]
dedupByAngleTwoA    = dedupByAngleTest  [(Point 3.0 2.0), (Point 1.0 2.0), (Point 2.0 4.0), (Point 1.5 1.0)]
                                        [(Point 3.0 2.0), (Point 2.0 4.0)]
dedupByAngleTwoB    = dedupByAngleTest  [(Point 1.0 2.0), (Point 2.0 4.0), (Point 3.0 2.0), (Point 1.5 1.0)]
                                        [(Point 2.0 4.0), (Point 3.0 2.0)]

-- Sorting tests
sortTest1 :: Assertion
sortTest1 = assertEqual "Empty points list did not result in an empty list" [] (sortPoints startPoint [])

sortError :: String
sortError = "Points were not sorted correctly"

t2In = [(Point 10.0 5.0), (Point 0.0 10.0), (Point 10.0 10.0), (Point 10.0 0.0), (Point 5.0 10.0)]
t2Out = [(Point 10.0 0.0), (Point 10.0 5.0), (Point 10.0 10.0), (Point 5.0 10.0), (Point 0.0 10.0)]
sortTest2 :: Assertion
sortTest2 = assertEqual sortError t2Out (sortPoints startPoint t2In)

t3In = [(Point 5.0 1.0), (Point 3.0 3.0), (Point 1.0 4.0), (Point 2.0 6.0), (Point 8.0 7.0), (Point 6.0 6.0), (Point 4.0 2.0)]
t3Out = [(Point 5.0 1.0), (Point 4.0 2.0), (Point 8.0 7.0), (Point 6.0 6.0), (Point 2.0 6.0), (Point 1.0 4.0)]
sortTest3 = assertEqual sortError t3Out (sortPoints startPoint t3In)

main :: IO ()
main = defaultMainWithOpts
       [testCase "dist1" distTest1
       , testCase "dist2" distTest2
       , testCase "start1" startTest1
       , testCase "start2" startTest2
       , testCase "polar0" polarTest0Deg
       , testCase "polar30" polarTest30Deg
       , testCase "polar60" polarTest60Deg
       , testCase "polar90" polarTest90Deg
       , testCase "polar120" polarTest120Deg
       , testCase "polar150" polarTest150Deg
       , testCase "polar180" polarTest180Deg
       , testCase "polarLine" polarTestLine
       , testCase "dedupByEqNone" dedupByEqNone
       , testCase "dedupByEqOne" dedupByEqOne
       , testCase "dedupByEqTriple" dedupByEqTriple
       , testCase "dedupByEqTwo" dedupByEqTwo
       , testCase "dedupByAngleNone" dedupByAngleNone
       , testCase "dedupByAngleOne" dedupByAngleOne
       , testCase "dedupByAngleTriple" dedupByAngleTriple
       , testCase "dedupByAngleTwoA" dedupByAngleTwoA
       , testCase "dedupByAngleTwoB" dedupByAngleTwoB
       , testCase "sort1" sortTest1
       , testCase "sort2" sortTest2
       , testCase "sort3" sortTest3]
       mempty
       --, testCase "" 

File: fsharp/GrahamScan/GrahamScanTests/PointTests.fs
namespace GrahamScanTests

open NUnit.Framework

open GrahamScan.GrahamScan
open GrahamScan.Point

open System

[<TestFixture>]
type PointTests () =
    static member private StartPoint = Point(0.0, 0.0)
    static member private Epsilon = 10e-6

    static member private DistanceError = "Distance was not calculated correctly"
    static member private StartError = "Did not find the correct start point"
    static member private PolarError = "Polar angle was not calculated correctly"
    static member private AnglesNotEqual = "Points did not have the same angle relative to the start point"
    static member private DedupError = "Did not deduplicate list correctly"
    static member private SortError = "Points were not sorted correctly"

    [<TestCaseSource("DistanceCases")>]
    static member DistanceTest x1 y1 x2 y2 expected =
        let p1 = Point(x1, y1)
        let p2 = Point(x2, y2)
        let actual = Point.Distance p1 p2
        Assert.AreEqual(expected, actual, PointTests.DistanceError)

    static member DistanceCases =
        [|
            [| 1.0; 1.0; 1.0; 1.0; 0.0 |];
            [| 3.0; 0.0; 0.0; 4.0; 5.0 |]
        |]

    [<TestCaseSource("StartingPointCases")>]
    static member StartingPointTest expected ps =
        let actual = StartingPoint ps
        Assert.AreEqual(expected, actual, PointTests.StartError)

    static member StartingPointCases: obj [] [] =
        [|
            [| PointTests.StartPoint; [ PointTests.StartPoint ] |] ;
            [| Point(1.0, 1.0); [ Point(2.0, 2.0); Point(2.0, 1.0); Point(1.0, 2.0); Point(1.0, 1.0) ] |]
        |]

    [<TestCaseSource("PolarAngleCases")>]
    static member PolarAngleTest expected p =
        let actual = Point.AngleBetween PointTests.StartPoint p
        Assert.AreEqual(expected, actual, PointTests.Epsilon, PointTests.PolarError)

    static member PolarAngleCases: obj [] [] =
        [|
            [| 0.0; Point(1.0, 0.0) |];
            [| Math.PI / 6.0; Point(0.5 * sqrt 3.0, 0.5) |];
            [| Math.PI / 3.0; Point(0.5, 0.5 * sqrt 3.0) |];
            [| Math.PI / 2.0; Point(0.0, 1.0) |];
            [| Math.PI * 2.0 / 3.0; Point(-0.5, 0.5 * sqrt 3.0) |];
            [| Math.PI * 5.0 / 6.0; Point(-0.5 * sqrt 3.0, 0.5) |];
            [| Math.PI; Point(-1.0, 0.0) |]
        |]

    [<TestCaseSource("PolarInLineCases")>]
    static member PolarInLineTest p1 p2 =
        let a1 = Point.AngleBetween PointTests.StartPoint p1
        let a2 = Point.AngleBetween PointTests.StartPoint p2
        Assert.AreEqual(a1, a2, PointTests.Epsilon, PointTests.AnglesNotEqual)

    static member PolarInLineCases =
        [|
            [| Point(0.0, 5.0); Point(0.0, 10.0) |];
            [| Point(1.0, 2.0); Point(2.0, 4.0) |]
        |]

    static member CoordsToPoint =
        let f = fun (x, y) -> Point(x, y)
        Array.map f >> List.ofArray

    [<TestCaseSource("DedupByEqCases")>]
    static member DedupByEqTest expecteds actuals =
        let expected = PointTests.CoordsToPoint expecteds
        let actual   = PointTests.CoordsToPoint actuals |> DedupByEq
        CollectionAssert.AreEqual(expected, actual, PointTests.DedupError)

    static member DedupByEqCases =
        [|
            [|
                [| (1.0, 2.0); (1.0, 1.0) |];
                [| (1.0, 2.0); (1.0, 1.0) |]
            |];
            [|
                [| (1.0, 1.0) |];
                [| (1.0, 1.0); (1.0, 1.0) |]
            |];
            [|
                [| (1.0, 1.0) |];
                [| (1.0, 1.0); (1.0, 1.0); (1.0, 1.0) |]
            |];
            [|
                [| (1.0, 1.0); (2.0, 3.0) |];
                [| (1.0, 1.0); (2.0, 3.0); (2.0, 3.0); (1.0, 1.0) |]
            |];
        |]

    [<TestCaseSource("DedupByAngleCases")>]
    static member DedupByAngleTest expecteds actuals =
        let expected = PointTests.CoordsToPoint expecteds
        let actual   = PointTests.CoordsToPoint actuals |> DedupByAngle PointTests.StartPoint
        CollectionAssert.AreEqual(expected, actual, PointTests.DedupError)

    static member DedupByAngleCases =
        [|
            [|
                [| (1.0, 2.0); (1.0, 1.0) |];
                [| (1.0, 2.0); (1.0, 1.0) |]
            |];
            [|
                [| (1.0, 1.0) |];
                [| (1.0, 1.0); (1.0, 1.0) |]
            |];
            [|
                [| (10.0, 20.0) |];
                [| (1.0, 2.0); (10.0, 20.0) |]
            |];
            [|
                [| (100.0, 200.0) |];
                [| (1.0, 2.0); (10.0, 20.0); (100.0, 200.0) |]
            |];
            [|
                [| (2.0, 4.0); (3.0, 2.0) |];
                [| (1.0, 2.0); (2.0, 4.0); (3.0, 2.0); (1.5, 1.0) |]
            |];
        |]

    [<TestCaseSource("SortByAngleCases")>]
    static member SortByAngleTest expecteds actuals =
        let expected = PointTests.CoordsToPoint expecteds
        let actual   = PointTests.CoordsToPoint actuals |> SortPoints PointTests.StartPoint
        CollectionAssert.AreEqual(expected, actual, PointTests.SortError)

    static member SortByAngleCases =
        [|
            [|
                [||];
                [||]
            |];
            [|
                [| (10.0, 0.0); (10.0, 5.0); (10.0, 10.0); (5.0, 10.0); (0.0, 10.0) |];
                [| (10.0, 5.0); (0.0, 10.0); (10.0, 10.0); (10.0, 0.0); (5.0, 10.0) |]
            |];
            [|
                [| (5.0, 1.0); (4.0, 2.0); (8.0, 7.0); (6.0, 6.0); (2.0, 6.0); (1.0, 4.0) |];
                [| (5.0, 1.0); (3.0, 3.0); (1.0, 4.0); (2.0, 6.0); (8.0, 7.0); (6.0, 6.0); (4.0, 2.0) |]
            |]
        |]